addons/hr_contract_timesheet_holidays/models/resource_calender_leave.py

```diff
@@ -0,0 +1,53 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+from odoo.osv import expression
+
+
+class ResourceCalendarLeaves(models.Model):
+    _inherit = "resource.calendar.leaves"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        leaves = super().create(vals_list)
+        leaves.filtered(lambda l: not l.resource_id)._assign_timesheets_for_public_holiday()
+        return leaves
+
+    def write(self, vals):
+        leaves = super().write(vals)
+        fields_to_check = {'date_from', 'date_to', 'calendar_id'}
+        if any(field for field in fields_to_check if field in vals):
+            self._assign_timesheets_for_public_holiday()
+        return leaves
+
```

Comment:
We should handle it in the _generate_timesheets method instead of creating a new method.